pkg/certificate/signer.go

```diff
@@ -0,0 +1,528 @@
+	CAKeyFileName   = "ca.key"
+	CACertFileName  = "ca.crt"
+)
+
+type SignerConfig struct {
+	RestConfig *rest.Config
+	DynClient  dynamic.Interface
+	RestMapper meta.RESTMapper
+}
+
+type Signer interface {
+	Start(namespace string) error
+	Stop(namespace string)
+	IssueCA(ctx context.Context, namespace string) error
+}
+
+type signerImpl struct {
+	restMapper meta.RESTMapper
+	dynClient  dynamic.Interface
+	syncerMap  sync.Map
+	namespace  string
+	caTicker   *time.Ticker
+	caStopCh   chan struct{}
+}
+
+func NewSigner(config SignerConfig) (Signer, error) {
+	s := &signerImpl{}
+
+	var err error
+
+	if config.RestMapper == nil {
+		if s.restMapper, err = util.BuildRestMapper(config.RestConfig); err != nil {
+			return nil, errors.Wrap(err, "error building the REST mapper")
+		}
+	}
+
+	if config.DynClient == nil {
+		if s.dynClient, err = resource.NewDynamicClient(config.RestConfig); err != nil {
+			return nil, errors.Wrap(err, "error creating dynamic client")
+		}
+	}
+
+	return s, nil
+}
+
+func (s *signerImpl) Start(namespace string) error {
+	if _, exists := s.syncerMap.Load(namespace); exists {
+		return nil
+	}
+
+	secretSyncer, err := syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:            "Cert Signer",
+		SourceClient:    s.dynClient,
+		SourceNamespace: namespace,
+		RestMapper:      s.restMapper,
+		ResourceType:    &corev1.Secret{},
+		ShouldProcess: func(obj *unstructured.Unstructured, op syncer.Operation) bool {
+			return op != syncer.Delete && obj.GetLabels()[SigningRequestLabelKey] != "" && obj.GetAnnotations()[RequestSignedLabelKey] == ""
+		},
+		Transform: func(from runtime.Object, _ int, _ syncer.Operation) (runtime.Object, bool) {
+			secret := from.(*corev1.Secret)
+
+			err := s.signSecret(secret)
+			if err != nil {
+				logger.Errorf(err, "error signing Secret %q", secret.Name)
+			}
+
+			return secret, err != nil
+		},
+		Federator: federate.NewUpdateFederator(s.dynClient, s.restMapper, namespace,
+			func(oldObj *unstructured.Unstructured, newObj *unstructured.Unstructured) *unstructured.Unstructured {
+				existingSecret := resource.MustFromUnstructured(oldObj, &corev1.Secret{})
+				updatedSecret := resource.MustFromUnstructured(newObj, &corev1.Secret{})
+
+				if existingSecret.Annotations == nil {
+					existingSecret.Annotations = map[string]string{}
+				}
+
+				existingSecret.Annotations[RequestSignedLabelKey] = "true"
+				existingSecret.Data[TLSDataKey] = updatedSecret.Data[TLSDataKey]
+				existingSecret.Data[CADataKey] = updatedSecret.Data[CADataKey]
+
+				return resource.MustToUnstructured(existingSecret)
+			}),
+	})
+	if err != nil {
+		return errors.Wrap(err, "error creating resource syncer")
+	}
+
+	stopCh := make(chan struct{})
+
+	if err := secretSyncer.Start(stopCh); err != nil {
+		return errors.Wrap(err, "error starting resource syncer")
+	}
+
+	s.syncerMap.Store(namespace, stopCh)
+
+	// Start periodic CA check if not already running
+	s.startPeriodicCACheck(namespace)
```

Comment:
We can ensure the CA check is only started once by:

```suggestion
	if _, loaded := s.syncerMap.LoadOrStore(namespace, stopCh); loaded {
		s.startPeriodicCACheck(namespace)
	}
```